Added hid_send_output_report() function

mbcinergy opened the pull request:
Where hid_get_input_report() was already available for some time, hid_send_output_report() was still missing. 

We have used our own fork with this function added for a couple of years now, but believe it's better to add this to the main repository as well. 

The changes are tested for Windows and on Linux (hidraw and libusb). 

The functions for mac and netbsd are also added but not tested, since we currently have no hardware available for that. It would be great if somebody can test this function for those platforms as well, but the functions are very straightforward so I don't expect any issues with them.

mbcinergy committed:
Added hid_send_output_report for Windows

mbcinergy committed:
added hid_send_output_report for libusb, lnux, mac and netbsd

mbcinergy committed:
modified HIDIOCSINPUT to HIDIOCSOUTPUT

Youw commented:
So far I haven't encountered cases when you have to use `hid_send_output_report` specifically, instead of using `hid_write`.

Most implementations of `hid_write` would try to use interrupt pipe to send the data, at first, and would fallback to `"hid_send_output_report"` if there is no interrupt pipe for a specific device.

On some platforms, the implementation of `hid_write` is identical to suggested `hid_send_output_report`, and on others - should work identically.
Have you tried it?

mbcinergy commented:
> So far I haven't encountered cases when you have to use `hid_send_output_report` specifically, instead of using `hid_write`.
> 
> Most implementations of `hid_write` would try to use interrupt pipe to send the data if possible first, and would fallback to `"hid_send_output_report"` if there is no interrupt (output) pipe for a specific device.
> 
> On some platforms, the implementation of `hid_write` is identical to suggested `hid_send_output_report`, and on others - should work identically (but "better"). Have you tried it?

I've tried `hid_write`, but it does not work in this case. The issue is that we have a couple of devices which do have an interrupt output pipe but use the control pipe for some vendor specific reports. That's why the `hid_write` does not work out of the box for those devices. So we need a function which is able to specify the pipe we need to write to.

Youw commented on hidapi/hidapi.h:
```diff
@@ -445,6 +445,38 @@ extern "C" {
+				This function returns the actual number of bytes written and
+				-1 on error.
+		*/
+		int HID_API_EXPORT HID_API_CALL hid_send_output_report(hid_device* dev, const unsigned char* data, size_t length);
```
Lets add to the documentation something like "see also @ref hid_write".
And lets update the documentation for `hid_write`, specifically:
> hid_write() will send the data on the first OUT endpoint, if one exists. If it does not, it will send the data through the Control Endpoint (Endpoint 0).

Would be great to update and specify that the behavior is as (newly introduced) `hid_send_output_report` in case if interrupt pipe does not exist.

Youw commented on libusb/hid.c:
```diff
@@ -1638,6 +1638,36 @@ int HID_API_EXPORT hid_get_feature_report(hid_device *dev, unsigned char *data,
 	return res;
 }
 
+int HID_API_EXPORT hid_send_output_report(hid_device *dev, const unsigned char *data, size_t length)
```
would be great to avoid code duplication among this function and `hid_write`

Youw commented on windows/hid.c:
```diff
@@ -1313,6 +1315,17 @@ int HID_API_EXPORT HID_API_CALL hid_get_feature_report(hid_device *dev, unsigned
 
+int HID_API_EXPORT HID_API_CALL hid_send_output_report(hid_device* dev, const unsigned char* data, size_t length)
+{
+	BOOL res = HidD_SetOutputReport(dev->device_handle, (PVOID)data, length);
```
I'm afraid this is not enought

take a look at `hid_send_feature_report` and try to extract common aproach/implementation, so that the two function would behavio alike

Youw replied:
I'm not sure why I marked this as resolved - it is clearly not.
See here:
https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/hidsdi/nf-hidsdi-hidd_setoutputreport#remarks
and see `hid_send_feature_report` implementation.

mbcinergy replied:
This is fixed in our branch. I think something happened with the references when I rebased on the latest master branch using the update button from this pull request (probably should have done a merge instead of rebase?).

Youw replied:
There are inifinite ways one can mess up git history. And there is (almost) always ways to fix it. Give it a try.

mbcinergy committed:
PR review changes

- added `hid_send_output_report` reference in function description for `hid_write`
- added see `hid_write` reference in function description for `hid_send_output_report`
- modified `hid_send_output_report` for windows to be more in line with `hid_send_feature_report`
- `hid_write` for libusb now calls `hid_send_output_report` if there is no interrupt out endpoint to prevent code duplication

mbcinergy committed:
added missing integer cast at return value

JoergAtGithub commented:
Wouldn't it be better to read the `Endpoint Descriptor` to determine if the particular interface of the USB device has a `Interupt Out`?

Youw commented:
Not sure what do you mean.
Right now libusb backend looks for a first interrupt endpoint of a _particular interface_. But it reads from a USB config descriptor.

I didn't check the linux/hidraw implementation, but Windows and macOS implementation is not exactly known.

JoergAtGithub commented:
I meant this: https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/standard-usb-descriptors#usb-endpoint-descriptor

mbcinergy committed:
added HIDIOCSOUTPUT define for kernels headers < 5.11

mbcinergy committed:
PR fix commit

Youw commented:
> I meant this: https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/standard-usb-descriptors#usb-endpoint-descriptor

I'm pretty sure that's what we do on libusb backend.
On other backends we don't have any control, as functions like `write` are the OS/Platform implementation, which appear to have in a similar way.
